oeps/oep-0045-arch-ops-and-config.rst

```diff
@@ -0,0 +1,175 @@
+* Similarly, bash scripts should be avoided.
+* If scripts must be used due to a limitation of the Dockerfile commands, the scripts must live within the same codebase as the Dockerfile.
+* Default ARG and CMD values should meet the needs of most users without modification.
+* Private or custom install requirements, patches, ARGs, CMD values, etc. should be included via a separately managed Dockerfile built on top of the Open edX image for that codebase.
+
+
+.. _native Dockerfile syntax: https://docs.docker.com/engine/reference/builder/#dockerfile-reference
+
+**Docker Images**
+
+edX will provide Docker images for applications that captures the latest code on the master branch as well as images representing named releases. edX will not provide these images for named releases prior to the acceptance and implementation of this OEP (Aspen through and including Juniper at time of writing).
+
+
+Configuration
+*************
+ 
+**Django settings**
+
+Having a single artifact that runs with different configurations increases stability by improving development parity with other deployment environments. edX applications already support configuration overrides via a yaml file for production environments, but development and test environments tend to configure the application using different code paths via a settings/devstack.py or settings/test.py file.
+
+Additionally it is not clear which settings are required to be overridden and which settings have values that may technically work but are inappropriate for production systems. To alleviate these issues edX Django applications will adopt the following settings structure:
+
+.. code-block:: text
+
+  settings
+  ├── __init__.py
+  ├── required.py
+  └── defaults.py
+
+
+*  ``__init__.py`` - Sourcing our config from this file within the settings directory takes advantage of Django defaults and means that settings will be picked up automatically without needing to specify ``--settings`` anywhere. This entry point would import ``required.py``, ``defaults.py``, and the code to override both from a config file.
+*  ``required.py`` - all settings which are required to run and do not have a reasonable production-ready default, e.g. LMS_BASE_URL which will be different per environment.
+*  ``defaults.py`` - other settings which will have production-ready defaults
+
+The settings defined in ``required.py`` and ``defaults.py`` files are mutually exclusive, representing all application specific settings as well as installed library settings whose values either must be provided or whose defaults are not considered production-ready.
+
+``required.py`` variables must be overridden by operators.  The application will check that operators provided these values, and will not start unless they are set. This allows operators to fail fast rather than finding out about an unset value when users exercise those breaking codepaths. Application developers are encouraged to keep the list of required settings to a minimum.
```

Comment:
If we're we're overriding the defaults from the config file, then I think you won't actually need to do any assignment in the `required.py`  it will only have the validation checks.  From what I can tell about sphinx autodoc, it looks like it's setup to pickup assignments, so you may end up in a situation where you have to do an assignment for the docs to pick it up even though it won't be necessary for the code to work correctly.

Reply:
It's not clear to my that `required.py` will do any validation, in which case it would exist entirely for documentation. I'm not sure how I feel about that. Could we instead just put required settings in `defaults.py`, and used Sphinx annotations to indicate that they're required.

If I'm wrong, and `required.py` will in fact do validation, then I recommend renaming `required.py` to `validate_required.py`, and clarifying that it will be imported *after* the YAML is loaded.

Reply:
Because the implementation details of this part matter quite a bit to the final design, I think I would like to move this conversation over to the "how we get there" part.  We should be open to amending this section as we coalesce around exactly what operating pattern needs to exist to realistically land somewhere where we have 1 codepath for configuration and clear delineations for operators around what they must provide to run the app.